File: mpi/code/fortran/solution/helloworld_mpi_b.f90
program helloworld_mpi_b

  use mpi

  implicit none

  integer                                       :: nbTask, myRank, ierr
  character(len=MPI_MAX_PROCESSOR_NAME)         :: proc_name
  integer                                       :: string_length
  character(len=MPI_MAX_LIBRARY_VERSION_STRING) :: mpi_library_name
  integer                                       :: mpi_version_num
  integer                                       :: mpi_subversion_num

  call MPI_Init(ierr)

  call MPI_COMM_SIZE(MPI_COMM_WORLD, nbTask, ierr)
  call MPI_COMM_RANK(MPI_COMM_WORLD, myRank, ierr)
  call MPI_GET_PROCESSOR_NAME(proc_name, string_length, ierr)

  if (myRank == 0) then
     call MPI_GET_VERSION(mpi_version_num, mpi_subversion_num,ierr)
     write (*,'("MPI standard version is ",i1,".",i1)') , mpi_version_num,mpi_subversion_num
     write (*,"()") 

     ! this is a MPI-3 routine
     ! you must have openmpi 1.8.x at least
     call MPI_GET_LIBRARY_VERSION(mpi_library_name, string_length, ierr)
     write (*,*) 'MPI implementation is:',mpi_library_name

  end if

  write (*,*) 'I am task', myRank, 'out of',nbTask, ' on processor ',proc_name

  call MPI_Finalize(ierr)

end program helloworld_mpi_b

File: mpi/code/c/solution/helloworld_mpi_b.c
#include <stdlib.h>
#include <stdio.h>

#include <mpi.h>

int main(int argc, char* argv[])
{

  int nbTask;
  int myRank;

  char proc_name[MPI_MAX_PROCESSOR_NAME];
  char mpi_library_name[MPI_MAX_LIBRARY_VERSION_STRING];
  int string_length;

  int mpi_version, mpi_subversion;

  MPI_Init(&argc, &argv);

  MPI_Comm_size(MPI_COMM_WORLD, &nbTask);
  MPI_Comm_rank(MPI_COMM_WORLD, &myRank);

  if (myRank == 0) {
    MPI_Get_version(&mpi_version, &mpi_subversion);
    printf("MPI standard version is %d.%d\n\n",mpi_version,mpi_subversion);

    // this is a MPI-3 routine
    // you must have openmpi 1.8.x at least
    MPI_Get_library_version(mpi_library_name, &string_length);
    printf("MPI implementation is:\n%s\n",mpi_library_name);

  }

  MPI_Get_processor_name(proc_name, &string_length);

  printf("I am task %d out of %d on processor %s\n",myRank,nbTask,proc_name);

  MPI_Finalize();

  return 0;

}
